### PR TITLE
Add emoji format to fingerprint commands

### LINKED ISSUE #570: certificate/fingerprint: emoji sum
### What would you like to be added

I would like to print the [sha256 fingerprint](https://smallstep.com/docs/step-cli/reference/certificate/fingerprint) of a certificate encoded as emojis with the help of https://github.com/emojisum/emojisum

### Why this is needed

Visually comparing the hex encoded sha256 is very cumbersome and error prone.
I think encoding it in emojis is a usability boost. Plus it is cute.

### Alternatives

The alternative is to simply pipe it to `emojisum` binary, which is doable, but requires yet another tool.
Of course, I also understand if you don't want to add the additional dependency.

### PR DESCRIPTION
**Pull request by oncilla**
Add an additional `emoji` format for the `step certificate fingerprint`
that encodes the certificate fingerprint as an emoji string.

Add the `--format` flag to the `step ssh fingerprint` command that
controls the encoding of the fingerprint. The default behavior of the
command is unchanged. The additional encodings are the same as for
`step certificate fingerprint`.

Add the `--format` flag to the `step crypto key fingerprint` command that
controls the encoding of the fingerprint. The default behavior of the
command is unchanged. For `sha1` hex encoding is used, for `sha256`
base64 encoding is used. The additional encodings are the same as for
`step certificate fingerprint`.

fixes #570

### CODE REVIEW

**Comment by maraino** [comment]
@oncilla I haven't looked at your changes yet, but what do you think about just creating the [256]rune with the mapping instead of importing the library.

**Comment by oncilla** [comment]
@maraino good idea. done.

**Comment by maraino** [comment]
@oncilla emojisum has some issues, see the issue I posted https://github.com/emojisum/emojisum/issues/37

But I kind of fixed it if we use the emojiCodeMap from my gist:
https://gist.github.com/maraino/5a15ced81d5467cd408ee07adbe88eb4

It might be preferable to use the codemap instead of the emojis to avoid conversion issues caused by an editor.

**Comment by dopey** [comment]
Hey @oncilla, let us know if need any more guidance here. cheers!

**Comment by oncilla** [comment]
@dopey should be okay, I just need to find some time to do the changes 🙂

**Comment by maraino** [comment]
Make sure to use the `emojiCodeMap` from my gist.

**Comment by oncilla** [comment]
@maraino @dopey should be ready now :+1:

**Comment by dopey** [comment]
Hey @oncilla, wanted to reiterate how grateful we are for the PR and your willingness to work through our comments to get it merged :]

We're still reviewing, but in the mean time, we wanted to show our appreciation by sending you some swag. If you're interested, shoot me an email (max@smallstep.com). Cheers!

File: command/command.go [hunk 1]
```diff
@@ -210,3 +211,33 @@ func setEnvVar(c *cli.Command) {
...
+	case "base64url", "base64-url":
+		return fingerprint.Base64URLFingerprint, nil
+	case "base64raw", "base64-raw":
+		return fingerprint.Base64RawFingerprint, nil
+	case "emoji", "emojisum":
+		return fingerprint.EmojiFingerprint, nil
+	default:
+		return 0, errors.Errorf("error parsing fingerprint format: '%s' is not a valid fingerprint format", format)
+	}
+}
```
Comment (maraino) [inline command/command.go:243]: base64.RawStdEncoding is missing, and is the default one for SSH fingerprints. See the [implementation].(https://github.com/golang/crypto/blob/ae814b36b87190c757eede9bc2d32ed77df88551/ssh/keys.go#L1479-L1483)

So `base64-raw` should be equivalent to base64.RawStdEncoding, and a new one of `base64-url-raw` should be base64.RawURLEncoding.

File: command/command.go [hunk 2]
```diff
@@ -210,3 +211,33 @@ func setEnvVar(c *cli.Command) {
 		setEnvVar(&c.Subcommands[i])
 	}
 }
+
+// FingerprintFormatFlag returns a flag for configuring the fingerprint format.
+func FingerprintFormatFlag() cli.StringFlag {
+	return cli.StringFlag{
+		Name:  "format",
+		Usage: `The <format> of the fingerprint, it must be "hex", "base64", "base64-url", "base64-raw", or "emoji".`,
+	}
```
Comment (maraino) [inline command/command.go:220]: You can also add a `Value: "hex"` here to use it as default, so you can get rid of the defaultFmt in the next method.

But, you will need a second flag with `base64-raw` for `step ssh fingerprint` as the default one is base64.RawStdEncoding. See next comment.

**Review by maraino — CHANGES_REQUESTED** [review]
Pretty much ok, but base64.RawStdEncoding is missing.

File: crypto/fingerprint/emoji.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+	for _, r := range input {
+		b.WriteString(emojiCodeMap[r])
+	}
+	return b.String()
+}
+
+// emojiCodeMap is a mapping from byte to emoji.
+//
+// The mapping is based on https://github.com/emojisum/emojisum with amendments
+// proposed in https://github.com/emojisum/emojisum/issues/37.
```
Comment (vbatts) [inline crypto/fingerprint/emoji.go:16]: fixed! feel free to pull it in.
I tagged a release `draft+2` for easier referencing. https://github.com/emojisum/emojisum/releases/tag/draft%2B2

**Comment by dopey** [comment]
Hey @oncilla let us know if you have any questions about the requested changes. cheers 🍻

**Comment by oncilla** [comment]
Hi @dopey

Sorry, Life got in the way. I will try to address the comments this week.

I presume you still want to add the emoji map in the code and not add emojisum as a dependency, right?

**Comment by dopey** [comment]
No worries at all @oncilla (and definitely no rush - just checking in). And yes, I think we want to store the map internally to the code base.

File: crypto/sshutil/fingerprint.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		comment = "no comment"
+	}
+
+	typ, size, err := publicKeyTypeAndSize(key)
+	if err != nil {
+		return "", errors.Wrap(err, "error determining key type and size")
+	}
+
+	fp := ssh.FingerprintSHA256(key)
+	if o.FingerprintOptions != nil {
```
[the remark this answers is not in the document]
Comment (oncilla) [reply crypto/sshutil/fingerprint.go]: Done.

File: command/ssh/fingerprint.go [hunk 5]
```diff
@@ -43,6 +47,23 @@ func fingerprint(ctx *cli.Context) error {
...
 	}
 
+	var (
+		format = ctx.String("format")
+
+		opts []libfingerprint.Option
+	)
+
+	if format != "" {
+		encoding, err := command.GetFingerprintEncoding(format, "hex")
```
[the remark this answers is not in the document]
Comment (oncilla) [reply command/ssh/fingerprint.go]: Done.

File: crypto/fingerprint/fingerprint.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+func encode(input []byte, encoding Encoding) string {
+	switch encoding {
+	case HexFingerprint:
+		return strings.ToLower(hex.EncodeToString(input))
+	case Base64Fingerprint:
+		return base64.StdEncoding.EncodeToString(input)
+	case Base64URLFingerprint:
+		return base64.URLEncoding.EncodeToString(input)
+	case Base64RawFingerprint:
+		return base64.RawStdEncoding.EncodeToString(input)
```
[the remark this answers is not in the document]
Comment (oncilla) [reply crypto/fingerprint/fingerprint.go]: Done.

I did not change the x509util constants as that would imply a breaking change.

File: command/command.go (hunk 1 above)
Comment (oncilla) [reply command/command.go]: Done.

File: command/command.go (hunk 2 above)
Comment (oncilla) [reply command/command.go:219]: Done.

File: crypto/fingerprint/emoji.go (hunk 3 above)
Comment (oncilla) [reply crypto/fingerprint/emoji.go]: Done.

**Comment by oncilla** [comment]
@dopey found some time today :tada: 
All comments should be addressed now.

**Review by maraino — APPROVED** [review]
Thanks @oncilla it looks great.

**Pull request merged**